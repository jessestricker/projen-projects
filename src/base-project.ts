import {
  Project,
  ProjectOptions,
  javascript,
  typescript,
  github,
  SampleReadme,
} from "projen";

/**
 * Options for the base project.
 */
export interface BaseProjectOptions extends ProjectOptions {
  /**
   * The Node package manager used to install dependencies and run scripts.
   *
   * @default javascript.NodePackageManager.NPM
   */
  readonly packageManager?: javascript.NodePackageManager;

  /**
   * The NPM dependencies used for development.
   */
  readonly devDeps?: string[];

  /**
   * Enable the GitHub integration.
   *
   * Enabled by default for root projects. Disabled for non-root projects.
   *
   * @default true
   */
  readonly github?: boolean;

  /**
   * The options for the GitHub integration.
   *
   * @default - see GitHubOptions
   */
  readonly githubOptions?: github.GitHubOptions;
}

export class BaseProject extends Project {
  readonly github?: github.GitHub;

  private annotatedAsGenerated?: Set<string>;

  constructor(options: BaseProjectOptions) {
    if (options.projenrcJson || options.projenrcJsonOptions) {
      throw new Error(".projenrc.json not supported");
    }

    const packageManager =
      options.packageManager ?? javascript.NodePackageManager.NPM;
    const packageRunnerCommand = getPackageRunnerCommand(packageManager);

    super({
      ...options,
      projenCommand: `${packageRunnerCommand} projen`,
    });

    // node package
    const nodePackage = new javascript.NodePackage(this, {
      devDeps: options.devDeps,
      entrypoint: "",
      packageManager: packageManager,
      licensed: false,
    });
    nodePackage.addField("private", true);
    nodePackage.addField("name", undefined);
    nodePackage.addField("publishConfig", undefined);
    nodePackage.addField("version", undefined);

    nodePackage.addDevDeps("projen", "constructs@^10.0.0");

    // projenrc file
    const projenrc = new typescript.ProjenrcTs(this);
    nodePackage.addDevDeps("ts-node");
    this.defaultTask?.reset(
      `${packageRunnerCommand} ts-node --project ${projenrc.tsconfig.fileName} ${projenrc.filePath}`,
    );

    // github
    if (options.github ?? !this.parent) {
      this.github = new github.GitHub(this, options.githubOptions);
      new SampleReadme(this);
    }
  }

  override annotateGenerated(glob: string): void {
    if (this.annotatedAsGenerated === undefined) {
      this.annotatedAsGenerated = new Set<string>();
    }
    this.annotatedAsGenerated.add(glob);
  }

  override preSynthesize(): void {
    if (this.github) {
      this.annotatedAsGenerated?.forEach((glob) =>
        this.gitattributes.addAttributes(glob, "linguist-generated"),
      );
    }
  }
}

function getPackageRunnerCommand(
  packageManager: javascript.NodePackageManager,
) {
  switch (packageManager) {
    case javascript.NodePackageManager.YARN_BERRY:
      return "yarn dlx";
    case javascript.NodePackageManager.NPM:
      return "npx";
    case javascript.NodePackageManager.PNPM:
      return "pnpx";
    case javascript.NodePackageManager.BUN:
      return "bunx";
    default:
      throw new Error(`unsupported package manager "${packageManager}"`);
  }
}
