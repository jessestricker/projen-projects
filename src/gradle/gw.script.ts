import * as child_process from "node:child_process";
import * as fs from "node:fs";
import * as path from "node:path";

function main() {
  const scriptFileName =
    process.platform === "win32" ? "gradlew.bat" : "gradlew";

  const scriptFile = findFileUp(scriptFileName);
  if (scriptFile === undefined) {
    console.error("error: gradlew script file not found");
    process.exitCode = 1;
    return;
  }

  const scriptArgs = process.argv.slice(2);
  child_process.spawnSync(scriptFile, scriptArgs, {
    stdio: "inherit",
    shell: true,
  });
}

function findFileUp(name: string) {
  let searchDir = process.cwd();

  while (true) {
    const filePath = path.join(searchDir, name);
    if (isFile(filePath)) {
      return filePath;
    }

    const searchDirParent = path.dirname(searchDir);
    if (searchDirParent === searchDir) {
      return undefined;
    }

    searchDir = searchDirParent;
  }
}

function isFile(file: string) {
  const stats = fs.statSync(file, { throwIfNoEntry: false });
  return stats?.isFile() ?? false;
}

main();
