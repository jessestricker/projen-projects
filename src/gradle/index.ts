export * from "./gradle";
export * from "./gradle-project";
