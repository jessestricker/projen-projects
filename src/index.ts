export * from "./base-project";
export * from "./gradle";
