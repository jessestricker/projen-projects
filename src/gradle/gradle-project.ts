import * as path from "node:path";
import { Component, SampleFile } from "projen";
import { BaseProject, BaseProjectOptions } from "../base-project";
import { Gradle, GradleOptions } from "./gradle";

/**
 * The Gradle project options.
 */
export interface GradleProjectOptions extends BaseProjectOptions {
  /**
   * @default {}
   */
  readonly gradleOptions?: GradleOptions;

  /**
   * @default true
   */
  readonly sampleCode?: boolean;
}

/**
 * A Gradle project using the Kotlin DSL.
 */
export class GradleProject extends BaseProject {
  readonly gradle: Gradle;

  constructor(options: GradleProjectOptions) {
    super(options);

    this.gradle = new Gradle(this, options.gradleOptions ?? {});

    if (options.sampleCode ?? true) {
      new SampleCode(this);
    }
  }
}

class SampleCode extends Component {
  constructor(project: GradleProject) {
    super(project);

    new SampleFile(project, "settings.gradle.kts", {
      contents: [
        "plugins {",
        '    id("org.gradle.toolchains.foojay-resolver-convention") version "0.9.0"',
        "}",
        "",
        `rootProject.name = "${project.name}"`,
        "",
      ].join("\n"),
    });
    new SampleFile(project, "build.gradle.kts", {
      sourcePath: path.join(
        __dirname,
        "..",
        "..",
        "sample-assets",
        "gradle-project",
        "build.gradle.kts",
      ),
    });
    new SampleFile(project, "gradle/libs.versions.toml", {
      sourcePath: path.join(
        __dirname,
        "..",
        "..",
        "sample-assets",
        "gradle-project",
        "libs.versions.toml",
      ),
    });
    new SampleFile(project, "src/main/kotlin/Main.kt", {
      sourcePath: path.join(
        __dirname,
        "..",
        "..",
        "sample-assets",
        "gradle-project",
        "Main.kt",
      ),
    });
  }
}
