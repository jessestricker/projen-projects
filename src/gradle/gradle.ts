import { Component } from "projen";
import { exec } from "projen/lib/util";
import { GradleProject } from "./gradle-project";

export interface GradleOptions {
  /**
   * The Gradle version to use.
   * @default "8.14"
   */
  readonly version?: string;
}

export class Gradle extends Component {
  readonly version: string;

  constructor(project: GradleProject, options: GradleOptions) {
    super(project);

    this.version = options.version ?? "8.14";

    // git
    project.gitignore.addPatterns("/.gradle/", "/build/");
    project.gitattributes.addAttributes("/gradlew", "text", "eol=lf");
    project.gitattributes.addAttributes("/gradlew.bat", "text", "eol=crlf");

    // tasks
    const compileTask = project.addTask("compile:gradle", {
      description: "Assembles main classes.",
      exec: "gw classes",
    });
    project.compileTask.spawn(compileTask);

    const packageTask = project.addTask("package:gradle", {
      description: "Assembles the outputs of this project.",
      exec: "gw assemble",
    });
    project.packageTask.spawn(packageTask);

    const testTask = project.addTask("test:gradle", {
      description: "Runs all checks.",
      exec: "gw check",
    });
    project.testTask.spawn(testTask);

    const cleanTask = project.addTask("clean:gradle", {
      description: "Deletes the Gradle build directory.",
      exec: "gw clean",
    });
    project.cleanTask.spawn(cleanTask);
  }

  override postSynthesize(): void {
    this.project.logger.info(`Adding the Gradle Wrapper...`);

    const generateGradleWrapper = (opts: { useWrapper: boolean }) => {
      const program = opts.useWrapper ? "gw" : "gradle";
      exec(`${program} wrapper --gradle-version ${this.version}`, {
        cwd: this.project.outdir,
      });
    };

    if (this.project.initProject) {
      generateGradleWrapper({ useWrapper: false });
    } else {
      generateGradleWrapper({ useWrapper: true });
    }

    // run a second time to re-generate scripts files
    generateGradleWrapper({ useWrapper: true });
  }
}
