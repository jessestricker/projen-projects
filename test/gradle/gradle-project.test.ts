import { Testing } from "projen";
import { GradleProject } from "../../src/gradle";

describe("GradleProject", () => {
  test("matches snapshot", () => {
    const project = new GradleProject({
      name: "gradle-project",
      gradleOptions: {
        version: "8.14",
      },
    });

    const snapshot = Testing.synth(project);

    expect(snapshot).toMatchSnapshot();
  });
});
