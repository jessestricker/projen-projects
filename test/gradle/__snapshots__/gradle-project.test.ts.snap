// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GradleProject matches snapshot 1`] = `
{
  ".gitattributes": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

* text=auto eol=lf
/.gitattributes linguist-generated
/.github/workflows/pull-request-lint.yml linguist-generated
/.gitignore linguist-generated
/.projen/** linguist-generated
/.projen/deps.json linguist-generated
/.projen/files.json linguist-generated
/.projen/tasks.json linguist-generated
/gradlew text eol=lf
/gradlew.bat text eol=crlf
/package-lock.json linguist-generated
/package.json linguist-generated
/tsconfig.projen.json linguist-generated",
  ".github/workflows/pull-request-lint.yml": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".

name: pull-request-lint
on:
  pull_request_target:
    types:
      - labeled
      - opened
      - synchronize
      - reopened
      - ready_for_review
      - edited
  merge_group: {}
jobs:
  validate:
    name: Validate PR title
    runs-on: ubuntu-latest
    permissions:
      pull-requests: write
    if: (github.event_name == 'pull_request' || github.event_name == 'pull_request_target')
    steps:
      - uses: amannn/action-semantic-pull-request@v5.4.0
        env:
          GITHUB_TOKEN: \${{ secrets.GITHUB_TOKEN }}
        with:
          types: |-
            feat
            fix
            chore
          requireScope: false
",
  ".gitignore": "# ~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".
node_modules/
!/.gitattributes
!/.projen/tasks.json
!/.projen/deps.json
!/.projen/files.json
!/package.json
!/tsconfig.projen.json
!/.github/workflows/pull-request-lint.yml
/.gradle/
/.idea/
/.kotlin/
/build/
!/.projenrc.ts
",
  ".projen/deps.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
    "dependencies": [
      {
        "name": "constructs",
        "type": "build",
        "version": "^10.0.0",
      },
      {
        "name": "projen",
        "type": "build",
      },
      {
        "name": "ts-node",
        "type": "build",
      },
    ],
  },
  ".projen/files.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
    "files": [
      ".gitattributes",
      ".github/workflows/pull-request-lint.yml",
      ".gitignore",
      ".projen/deps.json",
      ".projen/files.json",
      ".projen/tasks.json",
      "tsconfig.projen.json",
    ],
  },
  ".projen/tasks.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
    "tasks": {
      "build": {
        "description": "Full release build",
        "name": "build",
        "steps": [
          {
            "spawn": "default",
          },
          {
            "spawn": "pre-compile",
          },
          {
            "spawn": "compile",
          },
          {
            "spawn": "post-compile",
          },
          {
            "spawn": "test",
          },
          {
            "spawn": "package",
          },
        ],
      },
      "clean": {
        "description": "Deletes the build output.",
        "name": "clean",
        "steps": [
          {
            "spawn": "clean:gradle",
          },
        ],
      },
      "clean:gradle": {
        "description": "Deletes the Gradle build directory.",
        "name": "clean:gradle",
        "steps": [
          {
            "exec": "gw clean",
          },
        ],
      },
      "compile": {
        "description": "Only compile",
        "name": "compile",
        "steps": [
          {
            "spawn": "compile:gradle",
          },
        ],
      },
      "compile:gradle": {
        "description": "Assembles main classes.",
        "name": "compile:gradle",
        "steps": [
          {
            "exec": "gw classes",
          },
        ],
      },
      "default": {
        "description": "Synthesize project files",
        "name": "default",
        "steps": [
          {
            "exec": "ts-node --project tsconfig.projen.json .projenrc.ts",
          },
        ],
      },
      "eject": {
        "description": "Remove projen from the project",
        "env": {
          "PROJEN_EJECTING": "true",
        },
        "name": "eject",
        "steps": [
          {
            "spawn": "default",
          },
        ],
      },
      "install": {
        "description": "Install project dependencies and update lockfile (non-frozen)",
        "name": "install",
        "steps": [
          {
            "exec": "npm install",
          },
        ],
      },
      "install:ci": {
        "description": "Install project dependencies using frozen lockfile",
        "name": "install:ci",
        "steps": [
          {
            "exec": "npm ci",
          },
        ],
      },
      "package": {
        "description": "Creates the distribution package",
        "name": "package",
        "steps": [
          {
            "spawn": "package:gradle",
          },
        ],
      },
      "package:gradle": {
        "description": "Assembles the outputs of this project.",
        "name": "package:gradle",
        "steps": [
          {
            "exec": "gw assemble",
          },
        ],
      },
      "post-compile": {
        "description": "Runs after successful compilation",
        "name": "post-compile",
      },
      "pre-compile": {
        "description": "Prepare the project for compilation",
        "name": "pre-compile",
      },
      "test": {
        "description": "Run tests",
        "name": "test",
        "steps": [
          {
            "spawn": "test:gradle",
          },
        ],
      },
      "test:gradle": {
        "description": "Runs all checks.",
        "name": "test:gradle",
        "steps": [
          {
            "exec": "gw check",
          },
        ],
      },
    },
  },
  "README.md": "# replace this",
  "build.gradle.kts": "plugins {
    alias(libs.plugins.kotlin.jvm)
}

repositories {
    mavenCentral()
}

kotlin {
    jvmToolchain(21)
}
",
  "gradle/libs.versions.toml": "[versions]
kotlin = "2.1.21"

[plugins]
kotlin-jvm = { id = "org.jetbrains.kotlin.jvm", version.ref = "kotlin" }
",
  "package.json": {
    "//": "~~ Generated by projen. To modify, edit .projenrc.ts and run "npx projen".",
    "devDependencies": {
      "constructs": "^10.0.0",
      "projen": "*",
      "ts-node": "*",
    },
    "license": "UNLICENSED",
    "private": true,
    "scripts": {
      "build": "npx projen build",
      "clean": "npx projen clean",
      "clean:gradle": "npx projen clean:gradle",
      "compile": "npx projen compile",
      "compile:gradle": "npx projen compile:gradle",
      "default": "npx projen default",
      "eject": "npx projen eject",
      "package": "npx projen package",
      "package:gradle": "npx projen package:gradle",
      "post-compile": "npx projen post-compile",
      "pre-compile": "npx projen pre-compile",
      "test": "npx projen test",
      "test:gradle": "npx projen test:gradle",
    },
  },
  "settings.gradle.kts": "plugins {
    id("org.gradle.toolchains.foojay-resolver-convention") version "0.9.0"
}

rootProject.name = "gradle-project"
",
  "src/main/kotlin/Main.kt": "fun main() {
    println("Hello, World!")
}
",
  "tsconfig.projen.json": {
    "compilerOptions": {},
    "exclude": [
      "node_modules",
    ],
    "include": [
      "**/*.ts",
      ".projenrc.ts",
      "projenrc/**/*.ts",
    ],
  },
}
`;
